proposal/color-4-new-spaces.md

```diff
@@ -0,0 +1,1984 @@
+
+* If the keyword argument `$space` is specified in `$args`:
+
+    * Let `space` be the result of [looking up a known color space] named
+      `$space`.
+
+    * If `space` is not a [known color space], and `space != origin-space`,
+      throw an error.
+
+      > We can make adjustments in an unknown space, but we can't do conversion.
+
+    * Let `color` be the result of [converting] `$color` to `space`.
+
+* Otherwise:
+
+  * Let `space` be `origin-space`.
+
+  * Let `color` be the value of `$color`.
+
+* Let `alpha` be `color`'s alpha property.
+
+* If the keyword argument `$alpha` is specified in `$args`:
+
+  * If `$alpha` is not a number, throw an error.
+
+  * Let `new-alpha` be the result of [percent-converting] `$alpha` with a max
+    of 1.
+
+  * Set `alpha` to the value of `new-alpha + alpha` clamped between 0 and 1.
+
+* Let `channel-args` be the remaining keyword arguments in `$args`, not
+  including `$space` or `$alpha` arguments.
+
+* Let `channels` be a list of the `color`'s channels.
+
+* For each keyword `key` and value `adjust` in `channel-args`:
+
+  * If `key` is a string in the format `channel<integer>`, set `key` to the
+    value of `<integer>`.
+
+    > This allows e.g. `color.adjust($color, $channel1: 0.25)` for adjusting
+    > color channels in unknown color spaces.
+
+  * If `key` is not the name or index of a channel in `channels`, throw an
+    error.
+
+  * Set `adjust` to the result of [validating] `adjust` as `key` channel in
+    `space`.
+
+    > This ensures that `adjust` is percent-converted when necessary, and
+    > throws errors when the adjustment is not a valid value for the space.
+
+  * Set the `channel` with name or index of `key` in `channels` to
+    `channel + adjust`, treating values of `none` as `0`.
+
+* Let `new` be a color in color space `space`, with `channels` channels, and an
+  alpha of `alpha`.
+
+* Return the result of [converting] `new` into `origin-space`.
+
+### `color.scale()`
+
+```
+scale($color, $args...)
+```
+
+This function is also available as a global function named `scale-color()`.
+
+* If `$color` is not a color in a [known color space], throw an error.
+
+* If any item in `$args` is not a keyword argument, throw an error.
+
+* Let `origin-space` be `$color`'s color space.
+
+* If the keyword argument `$space` is specified in `$args`:
+
+    * Let `space` be the result of [looking up a known color space] named
+      `$space`.
+
+    * If `space` is not a [known color space], throw an error.
+
+      > We cannot scale channels in an unknown space, since there are no
+      > defined boundaries for a given channel to scale in relation to.
+
+    * Let `color` be the result of [converting] `$color` to `space`.
+
+* Otherwise:
+
+  * Let `space` be `origin-space`.
+
+  * Let `color` be the value of `$color`.
+
+* Let `alpha` be `color`'s alpha property.
+
+* If the keyword argument `$alpha` is specified in `$args`:
+
+    * If `alpha == none`, throw an error.
```

Comment:
Why not treat it as zero? Or potentially just keep it as `none`?

Reply:
I don't think CSS Colors 4 defines a valid usage of `none` for the alpha value. IIRC, it only has `none` for color space channels, not for the alpha.

Reply:
The current spec does allow alpha components to be set to `none` - see eg https://drafts.csswg.org/css-color-4/#missing:

> All color functions allow any of their components to be specified as [none](https://drafts.csswg.org/css-color-4/#valdef-color-none).

and https://drafts.csswg.org/css-color-4/#funcdef-rgb for explicit mention in a color function.

But yes, I think we can treat `none` as `0` in this case. Making that change.